ssl/quic/quic_fifd.c

```diff
@@ -0,0 +1,202 @@
+    fifd->txpim                 = txpim;
+    fifd->get_sstream_by_id     = get_sstream_by_id;
+    fifd->get_sstream_by_id_arg = get_sstream_by_id_arg;
+    fifd->regen_frame           = regen_frame;
+    fifd->regen_frame_arg       = regen_frame_arg;
+    return 1;
+}
+
+void ossl_quic_fifd_cleanup(QUIC_FIFD *fifd)
+{
+    /* No-op. */
+}
+
+static void on_acked(void *arg)
+{
+    QUIC_TXPIM_PKT *pkt = arg;
+    QUIC_FIFD *fifd = pkt->fifd;
+    const QUIC_TXPIM_CHUNK *chunks = ossl_quic_txpim_pkt_get_chunks(pkt);
+    size_t i, num_chunks = ossl_quic_txpim_pkt_get_num_chunks(pkt);
+    QUIC_SSTREAM *sstream;
+    QUIC_CFQ_ITEM *cfq_item, *cfq_item_next;
+
+    /* STREAM and CRYPTO stream chunks, FINs and stream FC frames */
+    for (i = 0; i < num_chunks; ++i) {
+        sstream = fifd->get_sstream_by_id(chunks[i].stream_id,
+                                          fifd->get_sstream_by_id_arg);
+        if (sstream == NULL)
+            continue;
+
+        if (chunks[i].end >= chunks[i].start)
+            ossl_quic_sstream_mark_acked(sstream,
+                                         chunks[i].start, chunks[i].end);
+
+        if (chunks[i].has_fin && chunks[i].stream_id != UINT64_MAX)
+            ossl_quic_sstream_mark_acked_fin(sstream);
+    }
+
+    /* GCR */
+    for (cfq_item = pkt->retx_head; cfq_item != NULL; cfq_item = cfq_item_next) {
+        cfq_item_next = cfq_item->pkt_next;
+        ossl_quic_cfq_release(fifd->cfq, cfq_item);
+    }
+
+    ossl_quic_txpim_pkt_release(fifd->txpim, pkt);
+}
+
+static void on_lost(void *arg)
+{
+    QUIC_TXPIM_PKT *pkt = arg;
+    QUIC_FIFD *fifd = pkt->fifd;
+    const QUIC_TXPIM_CHUNK *chunks = ossl_quic_txpim_pkt_get_chunks(pkt);
+    size_t i, num_chunks = ossl_quic_txpim_pkt_get_num_chunks(pkt);
+    QUIC_SSTREAM *sstream;
+    QUIC_CFQ_ITEM *cfq_item, *cfq_item_next;
+
+    /* STREAM and CRYPTO stream chunks, FIN and stream FC frames */
+    for (i = 0; i < num_chunks; ++i) {
+        sstream = fifd->get_sstream_by_id(chunks[i].stream_id,
+                                          fifd->get_sstream_by_id_arg);
+        if (sstream == NULL)
+            continue;
+
+        if (chunks[i].end >= chunks[i].start)
+            ossl_quic_sstream_mark_lost(sstream,
+                                        chunks[i].start, chunks[i].end);
+
+        if (chunks[i].has_fin && chunks[i].stream_id != UINT64_MAX)
+            ossl_quic_sstream_mark_lost_fin(sstream);
+
+        /*
+         * Inform caller that stream needs an FC frame.
+         *
+         * Note: We could track whether an FC frame was sent originally for the
+         * stream to determine if it really needs to be regenerated or not.
+         * However, if loss has occurred, it's probably better to ensure the
+         * peer has up-to-date flow control data for the stream. Given that
+         * these frames are extremely small, we may as well always send it when
+         * handling loss.
```

Comment:
This should be handled by my TX packetiser implementation.

I suppose you could argue that it's suboptimal because an extra packet is needed, though since QUIC_SSTREAM returns the earliest data first the receiver will still be able to make headway once it receives the first retransmission packet.

It wouldn't be hard to adapt the FIFM to track if MAX_STREAM_DATA was sent, but I suspect it's not an urgent optimisation.